internal/cache/file/downloader/job.go

```diff
@@ -197,8 +203,105 @@ func (job *Job) updateFileInfoCache() (err error) {
 	return
 }
 
+// downloadObjectAsync downloads the backing GCS object into a file as part of
+// file cache using NewReader method of gcs.Bucket.
+//
+// Note: There can only be one async download running for a job at a time.
+// Acquires and releases LOCK(job.mu)
+func (job *Job) downloadObjectAsync() {
+	// Create and open cache file for writing object into it.
+	cacheFile, err := util.CreateFile(job.fileSpec, os.O_WRONLY)
+	if err != nil {
+		err = fmt.Errorf("downloadObjectAsync: error in creating cache file: %v", err)
+		job.failWhileDownloading(err)
+		return
+	}
+	defer func(cacheFile *os.File) {
+		err = cacheFile.Close()
+		if err != nil {
+			err = fmt.Errorf("downloadObjectAsync: error while closing cache file: %v", err)
+			job.failWhileDownloading(err)
+		}
+	}(cacheFile)
+
+	var newReader io.ReadCloser
+	var start, end, sequentialReadSize, newReaderLimit int64
+	end = int64(job.object.Size)
+	sequentialReadSize = int64(job.sequentialReadSizeMb) * util.MiB
+
+	for {
+		select {
+		case <-job.cancelCtx.Done():
+			return
+		default:
+			if start < end {
+				if newReader == nil {
+					newReaderLimit = min(start+sequentialReadSize, end)
+					newReader, err = job.bucket.NewReader(
+						job.cancelCtx,
+						&gcs.ReadObjectRequest{
+							Name:       job.object.Name,
+							Generation: job.object.Generation,
+							Range: &gcs.ByteRange{
+								Start: uint64(start),
+								Limit: uint64(newReaderLimit),
+							},
+							ReadCompressed: job.object.HasContentEncodingGzip(),
+						})
+					if err != nil {
+						err = fmt.Errorf(fmt.Sprintf("downloadObjectAsync: error in creating NewReader with start %d and limit %d: %v", start, newReaderLimit, err))
+						job.failWhileDownloading(err)
+						return
+					}
+				}
+
+				maxRead := min(end-start, ReadChunkSize)
+				_, err = cacheFile.Seek(start, 0)
+				if err != nil {
+					err = fmt.Errorf(fmt.Sprintf("downloadObjectAsync: error while seeking file handle, seek %d: %v", start, err))
+					job.failWhileDownloading(err)
+					return
+				}
+
+				// copy the contents from NewReader to cache file.
+				_, readErr := io.CopyN(cacheFile, newReader, maxRead)
+				if readErr != nil && readErr != io.EOF {
+					err = fmt.Errorf("downloadObjectAsync: error at the time of copying content to cache file %v", readErr)
+					job.failWhileDownloading(err)
+					return
+				}
+				start += maxRead
+				if readErr == io.EOF {
+					newReader = nil
+				}
+
+				job.mu.Lock()
+				job.status.Offset = start
+				err = job.updateFileInfoCache()
+				// Notify subscribers if file cache is updated.
+				if err == nil {
+					job.notifySubscribers()
+				}
+				job.mu.Unlock()
+				// change status of job in case of error while updating file cache.
+				if err != nil {
+					job.failWhileDownloading(err)
+					return
+				}
+			} else {
+				job.mu.Lock()
+				job.status.Name = COMPLETED
+				job.notifySubscribers()
```

Comment:
Should we notify everyone in case of job-completion? This is just to handle, if someone subscribes for offset greater than object-size.

Reply:
they will get error if they try to subscribe for offset greater than object size. That check is in Download method in next PR.